### DISCUSSION TITLE
use ensemble.update_weights to simulate a different temperature

### DISCUSSION THREAD

**Discussion by chenyubi14** [SSCHAcode/python-sscha#163] [Q&A] [answered]
Hi,

In the tutorial for second order phase transitions, it is claimed that update_weights can also simulate a different temperature. I am using first-principles methods to calculate ensemble which is very expensive, so I want to know how it is implemented.
```
# We must update the ensemble weights
# We can also use this function to simulate a different temperature.
ensemble.update_weights(dyn_sscha_final, Temperature)
```
Let's say I have two temperatures 100K and 300K. For 100K, I have minimized the free energy with 10 populations of 50 configurations, and I have obtained the 100K hessian with 5000 configurations. So far, I do not have any populations for 300K.

- Can I feed the 10 populations to 300K? I guess no, because I always need to minimize w.r.t. the new parameters.
- If I feed the 5000 configurations for 100K to 300K using `ensemble.update_weights`, do I need to run another 5000 configurations for 300K?

A detailed example for the work flow will be appreciated. 
Thanks!

**Comment by mesonepigreco** [comment, marked as answer]
Hi,
It is possible to reuse the ensemble to minimize at a different temperature, however, the temperature must be relatively close to the one already simulated (100 and 300 K are likely too different, try something like 150 K). 

To use it, just call the `update_weights` with the target temperature, and then run again the minimization. 
For example:

```
# Load the ensemble at 100 K that converged the calculation
dyn = CC.Phonons.Phonons(f"data/dyn_gen_pop{POPULATION}", nqirr)
ensemble = sscha.Ensemble.Ensemble(dyn, 100)
ensemble.load_bin("data_dir", POPULATION)

# Change the temperature of the ensemble to the desired one (for example 150 K)
ensemble.update_weights(dyn, 150)

# Initialize the minimizer and perform the free energy minimization
minim = sscha.SchaMinimizer.SSCHA_Minimizer(ensemble)
minim.init()
minim.run()

# Evaluate the free energy hessian
hessian = minim.ensemble.get_free_energy_hessian()
hessian.save_qe("hessian_at_150K")
```

Indeed, you can enrich the script with the usual tricks to print frequencies and minimization data or to change the convergence parameters. It could be a good idea to reduce the Kong-Liu ratio for the stochastic criterion to avoid an early exit as reweighting on temperature reduces the stochastic accuracy. 

In this way, you can get the hessian at temperatures near the one you simulated without any extra ab initio simulation.

Hope this helps!
All the best,
Lorenzo

**Reply by Kurufinve** [reply]
What about the case when we have calculated an ensemble at high temperature (e.g. 1000 K): can we reuse it for minimization at much lower temperatures (300 K for example)?

**Reply by mesonepigreco** [reply]
1000 K to 300 K seems still quite extreme, maybe 1000 K to 800 K could work. You can try, though. My experience is that from high temperatures to lower temperatures reweighting works better than the other way around.
